### PR TITLE
Enable use of Shadow DOM

### PR DESCRIPTION
**Pull request by ilyakamens**

### CODE REVIEW

File: lib/component.js [hunk 1]
```diff
@@ -55,7 +56,8 @@ export default class Component extends WebComponent {
     Object.assign(this.state, newState);
 
     this.loop = MainLoop(this.state, this._render.bind(this), {create, diff, patch});
-    this.appendChild(this.loop.target);
+    this.el = this._config.useShadowDom ? this.createShadowRoot() : this;
```
Comment (tdumitrescu) [inline lib/component.js]: `this.getConfig('useShadowDom')`

File: lib/component.js [hunk 2]
```diff
@@ -68,6 +70,12 @@ export default class Component extends WebComponent {
     this.$panelParent && this.$panelParent.$panelChildren.delete(this);
   }
 
+  attributeChangedCallback() {
+    if (this === this.$panelRoot) {
```
Comment (tdumitrescu) [inline lib/component.js]: `if (this.isPanelRoot) {`

Comment (tdumitrescu) [inline lib/component.js]: probably also want to check `this.initialized`

File: test/test-component.js [hunk 3]
```diff
@@ -86,6 +86,27 @@ describe('Simple Component instance', function() {
...
+    beforeEach(function(done) {
+      el.setConfig('useShadowDom', true);
+      document.body.appendChild(el);
+      window.requestAnimationFrame(function() {
+        done();
+      });
+    });
+
+    it('creates and uses a shadow root', function() {
+      expect(el.el).not.to.equal(el);
```
Comment (tdumitrescu) [inline test/test-component.js]: this would be a good spot for: `expect(el.shadowRoot).to.be.ok();`

File: test/test-component.js [hunk 4]
```diff
@@ -86,6 +86,27 @@ describe('Simple Component instance', function() {
...
+    });
+
+    it('renders its template', function() {
+      expect(document.getElementsByClassName('foo')).to.have.lengthOf(0);
+      expect(el.children).to.have.lengthOf(0);
+      expect(el.shadowRoot).not.to.equal(null);
```
Comment (tdumitrescu) [inline test/test-component.js]: does this work for `undefined` etc?

File: test/test-component.js [hunk 5]
```diff
@@ -86,6 +86,27 @@ describe('Simple Component instance', function() {
...
+      });
+    });
+
+    it('creates and uses a shadow root', function() {
+      expect(el.el).not.to.equal(el);
+    });
+
+    it('renders its template', function() {
+      expect(document.getElementsByClassName('foo')).to.have.lengthOf(0);
+      expect(el.children).to.have.lengthOf(0);
```
Comment (tdumitrescu) [inline test/test-component.js]: `to.have.length(0)`

File: test/test-component.js [hunk 6]
```diff
@@ -86,6 +86,27 @@ describe('Simple Component instance', function() {
       });
     });
   });
+
+  context('when using shadow root', function() {
```
Comment (tdumitrescu) [inline test/test-component.js]: 'when using shadow dom'

File: test/test-component.js (hunk 4 above)
Comment (ilyakamens) [inline test/test-component.js]: yes, but i think using `ok` would be better

File: test/test-component.js (hunk 5 above)
Comment (ilyakamens) [inline test/test-component.js]: talked offline. `lengthOf` is the way to go

File: lib/component.js [hunk 7]
```diff
@@ -68,6 +71,31 @@ export default class Component extends WebComponent {
     this.$panelParent && this.$panelParent.$panelChildren.delete(this);
   }
 
+  attributeChangedCallback() {
+    if (this.initialized && this.isPanelRoot) {
+      this._applyStyles();
```
Comment (tdumitrescu) [inline lib/component.js]: probably don't want to call this unless the correct attribute has changed

File: test/fixtures/shadow-dom-app.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
+import { Component } from '../../lib';
+import h from 'virtual-dom/virtual-hyperscript';
+
+var ShadowDomApp = document.registerElement('shadow-dom-app', class extends Component {
+  get config() {
+    return {
+      css: 'color: blue;',
```
Comment (tdumitrescu) [inline test/fixtures/shadow-dom-app.js]: it seems like the `css` prop is useless unless `useShadowDom` is set - should we have some sort of runtime check that you're not trying to use it when `useShadowDom` is false?

File: test/test-component.js [hunk 9]
```diff
@@ -86,6 +86,41 @@ describe('Simple Component instance', function() {
...
+      expect(document.getElementsByClassName('foo')).to.have.lengthOf(0);
+      expect(el.children).to.have.lengthOf(0);
+      expect(el.shadowRoot.children[1].className).to.equal('foo');
+    });
+
+    it('applies the styles', function() {
+      expect(el.shadowRoot.children[0].innerHTML).to.equal('color: blue;');
+    });
+
+    context('when applying override styles', function() {
```
Comment (tdumitrescu) [inline test/test-component.js]: can we have one test case where the override is already set before `attachedCallback` fires? e.g. `<my-widget style-override="background: red">`

Comment (ilyakamens) [inline test/test-component.js]: added

**Comment by tdumitrescu** [comment]
merged https://github.com/mixpanel/panel/commit/c9abb0812ffed5ed047c788533edc68379f29880, merci!

**Pull request closed without merging**